Makes chameleon behavior an extension

mikomyazaki opened the pull request:
No player-facing changes that require a changelog.

Eventually, I want to add some additional chameleon items or things that provide chameleon behavior to their container (e.g an integrated electronics component). So I decided to see if I could tidy up the chameleon code, by stuffing it all into an extension. I'd also like to generalize it for any /obj, but this current implementation doesn't quite do that.

Putting it up to get some advice on a problem, which is that the verb `/atom/proc/change()` does not show up on items that have the extension. After that's solved it should work

PsiOmegaDelta commented on code/game/objects/items/weapons/cards_ids.dm:
```diff
@@ -135,26 +126,9 @@ var/const/NO_EMAG_ACT = -50
-	. = ..()
-	if(user.skill_check(SKILL_DEVICES,SKILL_ADEPT))
-		to_chat(user, SPAN_WARNING("This ID card has some form of non-standard modifications."))
+/obj/item/weapon/card/emag/New()
```
`/Initialize()`

PsiOmegaDelta commented on code/game/objects/items/weapons/cards_ids.dm:
```diff
@@ -135,26 +126,9 @@ var/const/NO_EMAG_ACT = -50
-	if(user.skill_check(SKILL_DEVICES,SKILL_ADEPT))
-		to_chat(user, SPAN_WARNING("This ID card has some form of non-standard modifications."))
+/obj/item/weapon/card/emag/New()
+	..()
```
`.= ..()`

PsiOmegaDelta commented on code/datums/extensions/chameleon.dm:
```diff
@@ -0,0 +1,181 @@
+	..()
+
+	if(!chameleon_choices)
+		chameleon_choices = generate_chameleon_choices(parent_type)
```
You sure that you don't want `holder.parent_type` here?
Also, a more generic setup might be:
```
/datum/extension/chameleon/New(datum/holder, base_type)
    ..()
    if(!chameleon_choices)
        chameleon_choices = generate_chameleon_choices(base_type ?? holder.parent_type)
    [...]
```
This way it defaults to the `parent_type` of the holder but allow objects with deeper subtyping to do `set_extension(src, /datum/extension/chameleon, /obj/item/clothing/suit)` if needed

PsiOmegaDelta commented on code/datums/extensions/chameleon.dm:
```diff
@@ -0,0 +1,181 @@
+		return
+	//this is necessary, unfortunately, as initial() does not play well with list vars
+	var/obj/item/copy = new newtype(null)
+	var/obj/item/clothing/C = src // Need a better way to do this, want it to be general for all /obj
```
Standard BYOND appearance vars can be copied simply by using `C.appearance = copy`

mikomyazaki commented:
Thanks for the review, @PsiOmegaDelta  - Do you have any thoughts about the problem I'm having getting the `change` verb to show up?

mikomyazaki commented:
All works now.

PsiOmegaDelta commented on code/datums/extensions/chameleon.dm:
```diff
@@ -0,0 +1,170 @@
+	base_type = /datum/extension/chameleon
+	var/item_flags = ITEM_FLAG_INVALID_FOR_CHAMELEON
+	flags = EXTENSION_FLAG_IMMEDIATE
+	var/list/chameleon_choices
```
Add
`var/static/list/chameleon_choices_by_type`

PsiOmegaDelta commented on code/datums/extensions/chameleon.dm:
```diff
@@ -0,0 +1,170 @@
+	..()
+
+	if(!chameleon_choices)
+		chameleon_choices = generate_chameleon_choices(base_type || holder.parent_type)
```
```
if(!chameleon_choices)
    var/chameleon_type = base_type || holder.parent_type
    chameleon_choices = LAZYACCESS(chameleon_choices_by_type, chameleon_type)
    if(!chameleon_choices)
        chameleon_choices = generate_chameleon_choices(chameleon_type)
        LAZYSET(chameleon_choices_by_type, chameleon_type, chameleon_choices)
```

PsiOmegaDelta reviewed (changes requested):
This proposal should allow chameleon lists to be shared between the same types of chameleon items, reducing list creations.

mikomyazaki commented:
> This proposal should allow chameleon lists to be shared between the same types of chameleon items, reducing list creations.

I've implemented this, and chameleon_choices + chameleon_choices_by_type are populated properly. However, when it comes time to do `picked in chameleon_choices`, no input box pops up. 

Why would that happen, unless `chameleon_choices` was empty, since picked doesn't specify a type.

PsiOmegaDelta commented on code/datums/extensions/chameleon.dm:
```diff
@@ -0,0 +1,175 @@
+	else
+		src.verbs -= /atom/proc/chameleon_appearance
+
+/datum/extension/chameleon/proc/change(picked in chameleon_choices)
```
This only works for verbs and while there are ways to setup things like you want it's probably easier to do:

```
/atom/proc/chameleon_appearance()
	set name = "Change Appearance"
	set desc = "Activate the holographic appearance changing module."
	set category = "Chameleon Items"
	set src in usr

	if(has_extension(src,/datum/extension/chameleon))
		var/datum/extension/chameleon/C = get_extension(src, /datum/extension/chameleon)
		C.change(usr)
	else
		src.verbs -= /atom/proc/chameleon_appearance

/datum/extension/chameleon/proc/change(mob/user)
	var/choice = input(user, "Select a new appearance", "Select appearance") as null|anything in chameleon_choices
	if(choice)
		disguise(chameleon_choices[choice ], user)

	// if(istype(src,/obj/item/clothing)) - Aren't (shouldn't) appearance changes as well as icon updates handled by the disguise/OnDisguise procs? Just remove these lines
	// var/obj/item/clothing/C = src
	// C.update_clothing_icon()
```

mikomyazaki replied:
I've added this.

mikomyazaki commented:
Everything is currently working for all types of item .

I had to change from using `C.appearance` in `disguise()` back to copying the individual vars. For some reason, it wouldn't copy icon stuff with appearance (although it worked okay with an earlier version), instead removing them, although I'm not sure why.

PsiOmegaDelta commented on code/datums/extensions/chameleon.dm:
```diff
@@ -0,0 +1,178 @@
+	OnDisguise()
+	return copy
+
+/datum/extension/chameleon/accessory/disguise(var/newtype, var/mob/user)
```
Why isn't this an override of `/datum/extension/chameleon/accessory/OnDisguise()`?

PsiOmegaDelta commented on code/datums/extensions/chameleon.dm:
```diff
@@ -0,0 +1,178 @@
+		C.sprite_sheets = copy.sprite_sheets.Copy()
+	//copying sprite_sheets_obj should be unnecessary as chameleon items are not refittable.
+
+	OnDisguise()
```
The copy should be provided to the  `/OnDisguise()`-proc instead and then be deleted instead of returned

PsiOmegaDelta commented on code/datums/extensions/chameleon.dm:
```diff
@@ -0,0 +1,178 @@
+/datum/extension/chameleon/proc/change(mob/user)
+	var/choice = input(user, "Select a new appearance", "Select appearance") as null|anything in chameleon_choices
+	if(choice)
+		disguise(chameleon_choices[choice], user)
```
Could add an `/OnChange(user, holder)` proc and call it at this point, that way the various `/change()` overrides can be changed to
```
/datum/extension/chameleon/backpack/OnChange(mob/user, atom/C)
	if(ismob(C.loc))
		var/mob/M = C.loc
		M.update_inv_back()
```
Also means that the icon updates are only called when the disguise is (probably) changed, rather than even if the user cancelled out

PsiOmegaDelta commented on code/datums/extensions/chameleon.dm:
```diff
@@ -0,0 +1,178 @@
+		P.impact_type = initial(copy_projectile.impact_type)
+	return P
+
+/datum/extension/chameleon/gun/OnDisguise(var/obj/item/weapon/gun/copy)
```
This proc-call would also be fixed by the proposed `/OnDisguise()`-changes above, since the copy isn't currently provided

mikomyazaki commented:
I think I've addressed all these review points.

Also tweaked:
- Changed names of some of the chameleon items so instead of 'Change jumpsuit Appearance' it will be properly capitalized as 'Change Jumpsuit Appearance' etc.
- Subtype of backpack will now have the proper choice list.

PsiOmegaDelta commented on code/datums/extensions/chameleon.dm:
```diff
@@ -0,0 +1,169 @@
+
+/datum/extension/chameleon/Destroy()
+	. = ..()
+	atom_holder.verbs -= /atom/proc/chameleon_appearance
```
`atom_holder = null`

mikomyazaki replied:
Done.

PsiOmegaDelta commented on code/datums/extensions/chameleon.dm:
```diff
@@ -0,0 +1,169 @@
+		C.item_state_slots = copy.item_state_slots.Copy()
+	if(copy.sprite_sheets)
+		C.sprite_sheets = copy.sprite_sheets.Copy()
+	//copying sprite_sheets_obj should be unnecessary as chameleon items are not refittable.
```
Code that is always applicable no matter the subtype should be kept in `/disguise/()`
`/OnDisguise()` is only for subtype specialization

mikomyazaki replied:
Done.

PsiOmegaDelta commented on code/datums/extensions/chameleon.dm:
```diff
@@ -0,0 +1,169 @@
+
+/datum/extension/chameleon/accessory/OnDisguise(var/obj/item/clothing/accessory/copy)
+	..()
+	if (!copy)
```
`copy` should never be null when `/OnDisguise()` is called

mikomyazaki replied:
Removed this and another similar bit.

PsiOmegaDelta commented on code/datums/extensions/chameleon.dm:
```diff
@@ -0,0 +1,169 @@
+		OnChange(user,src)
+
+/datum/extension/chameleon/proc/OnChange() //contains icon updates
+	if(istype(src,/obj/item/clothing))
```
This `src` is the `/datum/extension/chameleon`-datum so this will never pass

mikomyazaki replied:
Corrected.

PsiOmegaDelta commented on code/datums/extensions/chameleon.dm:
```diff
@@ -0,0 +1,169 @@
+		disguise(chameleon_choices[choice], user)
+		OnChange(user,src)
+
+/datum/extension/chameleon/proc/OnChange() //contains icon updates
```
```
/datum/extension/chameleon/proc/OnChange(mob/user, atom/C)
   if(istype(C))
        C.update_clothing_icon()
```

mikomyazaki replied:
Done.

PsiOmegaDelta commented on code/datums/extensions/chameleon.dm:
```diff
@@ -0,0 +1,169 @@
+	var/choice = input(user, "Select a new appearance", "Select appearance") as null|anything in chameleon_choices
+	if(choice)
+		disguise(chameleon_choices[choice], user)
+		OnChange(user,src)
```
`OnChange(user, holder)`

mikomyazaki replied:
Done.

PsiOmegaDelta commented on code/datums/extensions/chameleon.dm:
```diff
@@ -0,0 +1,169 @@
+/datum/extension/chameleon
```
Can remove many of the istype-checks by adding
`expected_type = /obj/item/clothing`

mikomyazaki replied:
Done + removed type checks.

PsiOmegaDelta commented on code/datums/extensions/chameleon.dm:
```diff
@@ -0,0 +1,169 @@
+/datum/extension/chameleon
+	base_type = /datum/extension/chameleon
+	var/item_flags = ITEM_FLAG_INVALID_FOR_CHAMELEON
```
Why is this var on the datum?
It's only applicable to `/obj/item` types

mikomyazaki replied:
Not sure why I did that. Stuck it on the individual items instead.

PsiOmegaDelta commented on code/datums/extensions/chameleon.dm:
```diff
@@ -0,0 +1,169 @@
+		var/obj/item/clothing/C = src
+		C.update_clothing_icon()
+
+/datum/extension/chameleon/backpack/OnChange(mob/user,var/obj/item/weapon/storage/backpack/C)
```
Also add
```
/datum/extension/chameleon/backpack
    expected_type = /obj/item/weapon/storage/backpack
```

mikomyazaki replied:
Done.

PsiOmegaDelta commented on code/datums/extensions/chameleon.dm:
```diff
@@ -0,0 +1,169 @@
+/datum/extension/chameleon/backpack/OnChange(mob/user,var/obj/item/weapon/storage/backpack/C)
+	if(ismob(C.loc))
+		user.update_inv_back()
+
```
```
/datum/extension/chameleon/headset
    expected_type = /obj/item/device/radio/headset
```

mikomyazaki replied:
Done.

PsiOmegaDelta commented on code/datums/extensions/chameleon.dm:
```diff
@@ -0,0 +1,169 @@
+/datum/extension/chameleon/headset/OnChange(mob/user,var/obj/item/device/radio/headset/C)
+	if(ismob(C.loc))
+		user.update_inv_ears()
+
```
```
/datum/extension/chameleon/gun
    expected_type = /obj/item/weapon/gun
```

mikomyazaki replied:
Done.

mikomyazaki commented:
Everything works. One thing that didn't work as I expected was line 34 of chameleon.dm with `	var/obj/item/C = atom_holder`

Originally I thought this should work with `var/obj/item/C = holder`, but for the case with the chameleon gun for some reason holder was null. I don't really understand how holder can ever be null or why it should only be null in the case of the chameleon gun.... especially since atom_holder isn't null and that is defined by holder.

mikomyazaki commented:
Dealt with travis problem. :+1:

mikomyazaki committed:
Makes chameleon-ness an extension.

PsiOmegaDelta commented on code/modules/clothing/chameleon.dm:
```diff
@@ -1,376 +1,126 @@
 
 /obj/item/device/radio/headset/chameleon
-	name = "radio headset"
+	name = "Radio Headset"
```
`name = "radio headset"`

PsiOmegaDelta commented on code/modules/clothing/chameleon.dm:
```diff
@@ -1,376 +1,126 @@
 
 /obj/item/clothing/gloves/chameleon
-	name = "black gloves"
+	name = "Gloves"
```
`name = "gloves"`

PsiOmegaDelta commented on code/modules/clothing/chameleon.dm:
```diff
@@ -1,376 +1,126 @@
 
 /obj/item/clothing/mask/chameleon
-	name = "gas mask"
+	name = "Mask"
```
`name = "mask"`

PsiOmegaDelta commented on code/modules/clothing/chameleon.dm:
```diff
@@ -1,376 +1,126 @@
 
 /obj/item/clothing/glasses/chameleon
-	name = "Optical Meson Scanner"
+	name = "Goggles"
```
`name = "goggles"`

PsiOmegaDelta commented on code/modules/clothing/chameleon.dm:
```diff
@@ -1,376 +1,126 @@
-//**********************
 /obj/item/weapon/storage/backpack/chameleon
-	name = "backpack"
+	name = "Backpack"
```
`name = "backpack"`

PsiOmegaDelta commented on code/modules/clothing/chameleon.dm:
```diff
@@ -1,376 +1,126 @@
-//*******************
 /obj/item/clothing/shoes/chameleon
-	name = "black shoes"
+	name = "Shoes"
```
`name = "shoes"`

PsiOmegaDelta commented on code/modules/clothing/chameleon.dm:
```diff
@@ -1,376 +1,126 @@
 
 /obj/item/clothing/suit/chameleon
-	name = "armor"
+	name = "Armor"
```
`name = "armor"`

PsiOmegaDelta commented on code/modules/clothing/chameleon.dm:
```diff
@@ -1,376 +1,126 @@
 
 /obj/item/clothing/head/chameleon
-	name = "grey cap"
+	name = "Cap"
```
`name = "cap"`

PsiOmegaDelta commented on code/modules/clothing/chameleon.dm:
```diff
@@ -1,376 +1,126 @@
-//starts off as a jumpsuit
 /obj/item/clothing/under/chameleon
-	name = "jumpsuit"
+	name = "Jumpsuit"
```
`name = "jumpsuit"`
Or `name = "\improper Jumpsuit"` if it's supposed to be a brand-name or something

PsiOmegaDelta commented on code/datums/extensions/chameleon.dm:
```diff
@@ -0,0 +1,169 @@
+	var/atom/atom_holder
+	var/chameleon_type = null //if we want an item to have chameleon changes not of its parent_type or base_type
+
+/datum/extension/chameleon/gun
```
Why is this part separate from the proc-overrides?

PsiOmegaDelta commented on code/datums/extensions/chameleon.dm:
```diff
@@ -0,0 +1,169 @@
+
+/datum/extension/chameleon/Destroy()
+	. = ..()
+	atom_holder = null
```
I was unclear, the `atom_holder = null` line should've been in addition to the `atom_holder.verbs -= ./proc/chameleon_appearance` line. That's on me

However, I'm unsure if removing the verb actually works due to you initializing a new verb rather than using `/atom/proc/chameleon_appearance` directly. I should test this.

mikomyazaki replied:
I tested this by adding in another verb with the job of removing `/atom/proc/chameleon_appearance` and it didn't work. So... hmm.

PsiOmegaDelta replied:
Then test
```
/datum/extension/chameleon
    [...]
    var/chameleon_verb

/datum/extension/chameleon/New()
    [...]
    // Creating a proc this way automatically adds it to the list of verbs
    chamelon_verb = new/atom/proc/chameleon_appearance(atom_holder, "Change [atom_holder.name] Appearance")

/datum/extension/chameleon/Destroy()
    [...]
    atom_holder.verbs -= chamelon_verb
```

mikomyazaki replied:
Added this + tested. All seems to be working.

PsiOmegaDelta commented on code/datums/extensions/chameleon.dm:
```diff
@@ -0,0 +1,169 @@
+	var/obj/item/clothing/C = holder
+	C.body_parts_covered = copy.body_parts_covered
+
+/datum/extension/chameleon/accessory/OnDisguise(var/obj/item/clothing/accessory/copy)
```
Should probably also call `..()`
I believe some accessories can be equipped, and not only attached, and thus will need the appropriate `body_parts_covered` flags

PsiOmegaDelta commented on code/datums/extensions/chameleon.dm:
```diff
@@ -0,0 +1,169 @@
+	OnDisguise(copy)
+	qdel(copy)
+
+/datum/extension/chameleon/proc/OnDisguise(var/obj/item/clothing/copy) //copies the relevant appearance vars
```
The expected type isn't necessarily `/obj/item/clothing`, see backpacks/emag for example
Needs an `istype()`-check here or, preferably, for the expected type to be limited to `/obj/item` (i.e. `/datum/extension/chameleon/proc/OnDisguise(var/obj/item/copy)`) and have a `/datum/extension/chameleon/clothing` subtype

mikomyazaki commented:
Review stuff addressed / replied to. Thanks!

PsiOmegaDelta commented on code/datums/extensions/chameleon.dm:
```diff
@@ -0,0 +1,181 @@
+
+/datum/extension/chameleon/clothing/OnDisguise(var/obj/item/clothing/copy) //copies the relevant appearance vars
+	var/obj/item/clothing/C = holder
+	C.body_parts_covered = copy.body_parts_covered
```
This var exists on `/obj/item` level and should thus probably be in the calling `/disguise()`-proc

mikomyazaki replied:
Moved it there. Thanks.

PsiOmegaDelta commented on code/datums/extensions/chameleon.dm:
```diff
@@ -0,0 +1,178 @@
+
+/datum/extension/chameleon/backpack/OnChange(mob/user,var/obj/item/weapon/storage/backpack/C)
+	if(ismob(C.loc))
+		user.update_inv_back()
```
Shouldn't all these be
```
var/user/U = C.loc
U.update_*()
```
?

mikomyazaki replied:
Yes. Fixed thanks.

PsiOmegaDelta commented on code/datums/extensions/chameleon.dm:
```diff
@@ -0,0 +1,181 @@
+
+/datum/extension/chameleon/headset/OnChange(mob/user,var/obj/item/device/radio/headset/C)
+	if(ismob(C.loc))
+		var/mob/living/carbon/human/M = C.loc
```
You've validated that `C.loc` is a mob but not that it's the human subtype. Doesn't matter in this case because `/update_inv_ears()` is a `/mob`-proc but it is something that should be consistent.

mikomyazaki replied:
Right... fix'd. Thanks.

Spookerton commented:
@mikomyazaki 
Mostly works, except:

guns change appearance but choke like
```
[12:06:00] Runtime in chameleon.dm,156: Undefined variable /datum/extension/chameleon/gun/var/flags_inv
  proc name: OnDisguise (/datum/extension/chameleon/gun/OnDisguise)
  usr: The unknown (/mob/living/carbon/human) ([0x300007f]) (the deck) (84,95,3) (/turf/simulated/floor/tiled/white) (spookerton)
  usr.loc: The deck (/turf/simulated/floor/tiled/white) ([0x1018243]) (84,95,3) (/area/crew_quarters/sleep/cryo)
  src: /datum/extension/chameleon/gun (/datum/extension/chameleon/gun)
  call stack:
  /datum/extension/chameleon/gun (/datum/extension/chameleon/gun): OnDisguise(the dispersive blaster (/obj/item/weapon/gun/energy/incendiary_laser))
  /datum/extension/chameleon/gun (/datum/extension/chameleon/gun): disguise(/obj/item/weapon/gun/energy/in... (/obj/item/weapon/gun/energy/incendiary_laser), the unknown (/mob/living/carbon/human))
  /datum/extension/chameleon/gun (/datum/extension/chameleon/gun): change(the unknown (/mob/living/carbon/human))
  the dispersive blaster (/obj/item/weapon/gun/energy/chameleon): Change chameleon gun Appearance()
```

emags don't work at all and only ever choke like
```
[12:03:02] Runtime in chameleon.dm,31: Cannot create objects of type null.
  proc name: disguise (/datum/extension/chameleon/proc/disguise)
  usr: The unknown (/mob/living/carbon/human) ([0x300007f]) (the deck) (84,96,3) (/turf/simulated/floor/tiled/white) (spookerton)
  usr.loc: The deck (/turf/simulated/floor/tiled/white) ([0x101830b]) (84,96,3) (/area/crew_quarters/sleep/cryo)
  src: /datum/extension/chameleon/ema... (/datum/extension/chameleon/emag)
  call stack:
  /datum/extension/chameleon/ema... (/datum/extension/chameleon/emag): disguise(null, the unknown (/mob/living/carbon/human))
  /datum/extension/chameleon/ema... (/datum/extension/chameleon/emag): change(the unknown (/mob/living/carbon/human))
  the cryptographic sequencer (/obj/item/weapon/card/emag): Change cryptographic sequencer Appearance()
```
